service/discovery/aws/compute.go

```diff
@@ -0,0 +1,243 @@
+	types2 "github.com/aws/aws-sdk-go-v2/service/lambda/types"
+	"github.com/aws/smithy-go"
+)
+
+// computeDiscovery handles the AWS API requests regarding the computing services (EC2 and Lambda)
+type computeDiscovery struct {
+	virtualMachineAPI EC2API
+	functionAPI       LambdaAPI
+	isDiscovering     bool
+	awsConfig         *Client
+}
+
+// EC2API describes the EC2 api interface (for mock testing)
+type EC2API interface {
+	DescribeInstances(ctx context.Context,
+		params *ec2.DescribeInstancesInput,
+		optFns ...func(options *ec2.Options)) (*ec2.DescribeInstancesOutput, error)
+}
+
+// LambdaAPI describes the lambda api interface (for mock testing)
+type LambdaAPI interface {
+	ListFunctions(ctx context.Context,
+		params *lambda.ListFunctionsInput, optFns ...func(*lambda.Options)) (*lambda.ListFunctionsOutput, error)
+}
+
+// newFromConfigEC2 holds ec2.NewFromConfig(...) allowing a test function to mock it
+var newFromConfigEC2 = ec2.NewFromConfig
+
+// newFromConfigLambda holds lambda.NewFromConfig(...) allowing a test function tp mock it
+var newFromConfigLambda = lambda.NewFromConfig
+
+// NewComputeDiscovery constructs a new awsS3Discovery initializing the s3-virtualMachineAPI and isDiscovering with true
+func NewComputeDiscovery(client *Client) discovery.Discoverer {
+	return &computeDiscovery{
+		virtualMachineAPI: newFromConfigEC2(client.Cfg),
+		functionAPI:       newFromConfigLambda(client.Cfg),
+		isDiscovering:     true,
+		awsConfig:         client,
+	}
+}
+
+// Name is the method implementation defined in the discovery.Discoverer interface
+func (d computeDiscovery) Name() string {
+	return "AWS Compute"
+}
+
+// List is the method implementation defined in the discovery.Discoverer interface
+func (d computeDiscovery) List() (resources []voc.IsCloudResource, err error) {
+	listOfVMs, err := d.discoverVirtualMachines()
+	if err != nil {
+		return
+	}
+	for _, machine := range listOfVMs {
+		resources = append(resources, &machine)
+	}
+
+	listOfFunctions, err := d.discoverFunctions()
+	if err != nil {
+		return
+	}
+	for _, function := range listOfFunctions {
+		resources = append(resources, &function)
+	}
+
+	return
+}
+
+// discoverVirtualMachines discovers all VMs (in the current region)
+func (d *computeDiscovery) discoverVirtualMachines() ([]voc.VirtualMachine, error) {
+	resp, err := d.virtualMachineAPI.DescribeInstances(context.TODO(), &ec2.DescribeInstancesInput{})
+	if err != nil {
+		var ae smithy.APIError
+		if errors.As(err, &ae) {
+			err = formatError(ae)
+		}
+		return nil, err
+	}
+	var resources []voc.VirtualMachine
+	for _, reservation := range resp.Reservations {
+		for i := range reservation.Instances {
+			vm := &reservation.Instances[i]
+			computeResource := &voc.Compute{
+				CloudResource: &voc.CloudResource{
+					ID:   d.getARNOfVM(vm),
+					Name: d.getNameOfVM(vm),
+					// TODO(all): -1 or 0 as default value when no creation time is available?
+					CreationTime: -1,
```

Comment:
`0` is probably easier to catch. Even better would be, if creation time would be an actual `*time.Time` so we can just set it to `nil`. However, this would require a change in the ontology generation.

Reply:
Yes, I agree with `*time.Time` as better option. It is also the type returned by AWS (most of the times), e.g. for launchTime. But for now I will set it to `0` - I also don't think someone could create an AWS virtual machine in 1970.